pkg/controller/awselb/aws_elb_controller.go

```diff
@@ -0,0 +1,313 @@
+	}
+
+	machineInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMachine,
+		UpdateFunc: c.updateMachine,
+		DeleteFunc: c.deleteMachine,
+	})
+	c.machinesLister = machineInformer.Lister()
+	c.machinesSynced = machineInformer.Informer().HasSynced
+
+	c.syncHandler = c.syncMachine
+	c.enqueueMachine = c.enqueue
+	c.logger = log.WithField("controller", controllerName)
+
+	return c
+}
+
+// Controller monitors master machines and adds to ELBs as appropriate.
+type Controller struct {
+	client     clustopclient.Interface
+	kubeClient kubeclientset.Interface
+
+	// To allow injection of syncMachine for testing.
+	syncHandler func(hKey string) error
+	// used for unit testing
+	enqueueMachine func(machine *capiv1.Machine)
+
+	// machinesLister is able to list/get machines and is populated by the shared informer passed to
+	// NewController.
+	machinesLister lister.MachineLister
+	// machinesSynced returns true if the machine shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	machinesSynced cache.InformerSynced
+
+	// Machines that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	logger log.FieldLogger
+}
+
+func (c *Controller) addMachine(obj interface{}) {
+	machine := obj.(*capiv1.Machine)
+	clustoplog.WithMachine(c.logger, machine).Debug("adding machine")
+	c.enqueueMachine(machine)
+}
+
+func (c *Controller) updateMachine(old, cur interface{}) {
+	oldMachine := old.(*capiv1.Machine)
+	curMachine := cur.(*capiv1.Machine)
+	clustoplog.WithMachine(c.logger, oldMachine).Infof("updating machine")
+	c.enqueueMachine(curMachine)
+}
+
+func (c *Controller) deleteMachine(obj interface{}) {
+	machine, ok := obj.(*capiv1.Machine)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+			return
+		}
+		machine, ok = tombstone.Obj.(*capiv1.Machine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Machine %#v", obj))
+			return
+		}
+	}
+	clustoplog.WithMachine(c.logger, machine).Infof("deleting machine")
+	c.enqueueMachine(machine)
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machines will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	log.Infof("Starting awselb controller")
+	defer log.Infof("Shutting down awselb controller")
+
+	if !controller.WaitForCacheSync("machine", stopCh, c.machinesSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(machine *capiv1.Machine) {
+
+	// We only need to add masters to ELBs at this time.
+	if !controller.MachineHasRole(machine, capicommon.MasterRole) {
```

Comment:
This check should be done before calling `enqueueMachine` so that a controller that has its `enqueue` function replaced retains the logic that the controller only acts on master machines.